### PR TITLE
default v2

### PR DESCRIPTION
**Pull request by davidmdm**

### CODE REVIEW

File: test/parsing.test.ts [hunk 1]
```diff
@@ -839,10 +833,21 @@ describe('Zod Parsing', () => {
...
         b: z.number().default(() => 42),
         c: z.boolean().default(true),
-        d: z.null().default(),
+        d: z.null().default(null),
       });
       assert.deepEqual(schema.parse({}), { a: 'hello', b: 42, c: true, d: null });
-      assert.ok((schema as any)[coercionTypeSymbol]);
+    });
+
+    it('should return nested defaults with default object', () => {
```
Comment (davidmdm) [inline test/parsing.test.ts:841]: I added this test which is the use case we wanted. However I can't get the typing to work

File: src/types.ts [hunk 2]
```diff
@@ -509,16 +446,45 @@ export class NullableType<T extends AnyType> extends Type<Infer<T> | null> imple
...
+}
+
+// type X = {
+//   a: DefaultableType<StringType>;
+//   b: DefaultableType<NumberType>;
+// };
+
+type DS = DefaultableType<StringType>;
+
+type R = DS extends Type<any> ? (DS extends Type<infer K> ? K : DS) : any;
```
Comment (davidmdm) [inline src/types.ts]: I can't seem to be able to infer the type of of a DefaultableType<StringType> (substitute StringType for any other Type class)... It just goes to any... any insight?

Comment (davidmdm) [reply src/types.ts]: @pierreis

Comment (pierreis) [reply src/types.ts]: I am not sure.
That is quite disturbing indeed.

Let me check.

Comment (davidmdm) [reply src/types.ts]: I managed to get type Infer to worth with DefaultTypes by making Infer more exhaustive:

```typescript
export type Infer<T extends AnyType> = T extends DefaultableType<any>
  ? T extends DefaultableType<infer Schema>
    ? Schema extends Type<infer K>
      ? Eval<K>
      : never
    : never
  : T extends Type<infer K>
  ? Eval<K>
  : never;
```

But I still can't get the DefaultType<ObjectType<Shape>> to type the correct defaultValue (default keys as optional)

Comment (pierreis) [reply src/types.ts]: Sounds good. Debugging advanced types in TypeScript is hellish...
I still have no clue.

Regardless, it is extremely strange to see inference working for `Optional` but not for `Defaultable`, considering that the two are defined almost identically.

Comment (davidmdm) [reply src/types.ts]: I've managed to get very close. However I now have the DeepPartialShape type become possibly infinitely deep, it keeps doing that sometimes when something gets more complicated in the type system. 

I find that typescript is extremely buggy once you go a little advanced... I'll keep at it for a while... But although we've managed to introduce DefaultableType I still can't type the keys of defaultValue of object as optional... Losing my mind haha...

Comment (davidmdm) [reply src/types.ts]: I am considering throwing away deepPartial... It has been a thorn in my side getting it to keep compiling

Comment (pierreis) [reply src/types.ts]: I cannot really see a use-case from my point of view, so I wouldn't mind.

Comment (davidmdm) [reply src/types.ts]: I pushed with everything working except for the fact that deepPartialShape will no longer compile. Gonna sit on this for a little as I mull it over.

**Pull request closed without merging**